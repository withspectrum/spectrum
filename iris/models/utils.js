const { db } = require('./db');

export const NEW_DOCUMENTS = db
  .row('old_val')
  .eq(null)
  .and(db.not(db.row('new_val').eq(null)));

export const listenToNewDocumentsIn = (table, cb) => {
  return (
    db
      .table(table)
      .changes({
        includeInitial: false,
      })
      // Filter to only include newly inserted messages in the changefeed
      .filter(NEW_DOCUMENTS)
      .run({ cursor: true }, (err, cursor) => {
        if (err) throw err;
        cursor.each((err, data) => {
          if (err) throw err;
          // Call the passed callback with the message directly
          cb(data.new_val);
        });
      })
  );
};

export const parseRange = timeframe => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: string) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(db.row('lastSeen').during(db.now().sub(current), db.now()))
    .count()
    .default(0)
    .run();
};

export const getGrowth = async (
  table: string,
  range: string,
  field: string,
  filter: ?mixed
) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};
