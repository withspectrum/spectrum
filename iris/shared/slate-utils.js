// @flow
/**
 * This file is shared between server and client.
 * ⚠️ DON'T PUT ANY NODE.JS OR BROWSER-SPECIFIC CODE IN HERE ⚠️
 *
 * Note: This uses Flow comment syntax so this whole file is actually valid JS without any transpilation
 * The reason I did that is because create-react-app doesn't transpile files outside the source folder,
 * so it chokes on the Flow syntax.
 * More info: https://flow.org/en/docs/types/comments/
 */
var Raw = require('slate/lib/serializers/raw').default;
var Plain = require('slate/lib/serializers/plain').default;

function toJSON(state /*: Object */) {
  return Raw.serialize(state, { terse: true });
}
function toState(json /*: Object */) {
  return Raw.deserialize(json, { terse: true });
}

function toPlainText(state /*: Object*/) {
  return Plain.serialize(state);
}
function fromPlainText(string /*: string*/) {
  return Plain.deserialize(string);
}

module.exports = {
  toJSON: toJSON,
  toState: toState,
  toPlainText: toPlainText,
  fromPlainText: fromPlainText,
};
