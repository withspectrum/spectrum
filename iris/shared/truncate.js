// @flow
/**
 * This file is shared between server and client.
 * ⚠️ DON'T PUT ANY NODE.JS OR BROWSER-SPECIFIC CODE IN HERE ⚠️
 *
 * Note: This uses Flow comment syntax so this whole file is actually valid JS without any transpilation
 * The reason I did that is because create-react-app doesn't transpile files outside the source folder,
 * so it chokes on the Flow syntax.
 * More info: https://flow.org/en/docs/types/comments/
 */

// Truncate a string nicely to a certain length
function truncate(str /*: string */, length /*: number */) {
  if (str.length <= length) {
    return str;
  }
  var subString = str.substr(0, length);
  return (
    subString.substr(0, subString.lastIndexOf(' ')).replace(/\n/, ' ') + '…'
  );
}

module.exports = truncate;
