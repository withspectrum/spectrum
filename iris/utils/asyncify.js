// @flow
// Turn a callback-based listener into an async iterator
const Queue = require('then-queue');

const asyncify = (
  listener: (...args: any[]) => any,
  onError?: Error => void
) => {
  let queues = [];
  listener((...args) => queues.map(q => q.push(...args)));

  return () => {
    const queue = new Queue();
    queues.push(queue);

    // I have no idea how to type generators...
    // $FlowFixMe
    async function* AsyncGenerator() {
      try {
        while (true) {
          yield await queue.pop();
        }
      } catch (err) {
        // Remove the error'ing queue from the queues array
        queues = queues.filter(q => q !== queue);
        onError && onError(err);
      }
    }

    return AsyncGenerator();
  };
};

export default asyncify;
