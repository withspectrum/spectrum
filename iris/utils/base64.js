// @flow

/**
 * Encode a string to base64 (using the Node built-in Buffer)
 *
 * Stolen from http://stackoverflow.com/a/38237610/2115623
 */
export const encode = (string: string) =>
  Buffer.from(string).toString('base64');

type Base64String = string;
/**
 * Decode a base64 string (using the Node built-in Buffer)
 *
 * Stolen from http://stackoverflow.com/a/38237610/2115623
 */
export const decode = (string?: Base64String) =>
  (string ? Buffer.from(string, 'base64').toString('ascii') : '');
