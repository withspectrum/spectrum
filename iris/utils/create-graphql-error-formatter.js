// @flow
import Raven from 'raven';
import { IsUserError } from './UserError';
import type { GraphQLError } from 'graphql';

const createGraphQLErrorFormatter = (req?: express$Request) => (
  error: GraphQLError
) => {
  console.log('error', error);
  const isUserError = error.originalError
    ? error.originalError[IsUserError]
    : false;
  let sentryId = 'ID only generated in production';
  if (!isUserError) {
    if (process.env.NODE_ENV === 'production') {
      sentryId = Raven.captureException(
        error,
        req && Raven.parsers.parseRequest(req)
      );
    }
  }
  return {
    message: isUserError ? error.message : `Internal server error: ${sentryId}`,
    // Hide the stack trace in production mode
    stack:
      !process.env.NODE_ENV === 'production' ? error.stack.split('\n') : null,
  };
};

export default createGraphQLErrorFormatter;
