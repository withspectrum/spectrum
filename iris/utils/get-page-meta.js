const { parse } = require('url');
const generateMetaInfo = require('../../shared/generate-meta-info');

// Don't even try if the path is /<value> any of these
// TODO: Longer, more complete blacklist here
const PATH_BLACKLIST = ['robots.txt', 'home', 'messages', 'notifications'];

type Meta = {
  title: string,
  description: string,
  extra: string,
};

export default (
  url: string,
  request: (query: string) => Promise
): Promise<Meta> => {
  const { pathname = '' } = parse(url);
  const [, first, second, third] = pathname.split('/');
  if (third || first.length === 0) return Promise.resolve(generateMetaInfo());
  let promise;

  switch (first) {
    case 'explore': {
      promise = Promise.resolve(generateMetaInfo({ type: 'explore' }));
      break;
    }
    case 'thread': {
      /**
       * Thread
       */
      promise = request(/* GraphQL */ `
        {
          thread(id: "${second}") {
            content {
              title
              body
            }
            type
            channel {
              name
              isPrivate
            }
          }
        }
      `).then(res => {
        const { thread: { type, content, channel } } = res.data;
        return generateMetaInfo({
          type: 'thread',
          data: {
            title: content.title,
            type,
            body: content.body,
            channelName: channel.name,
            privateChannel: channel.isPrivate,
          },
        });
      });
      break;
    }
    case 'users': {
      /**
       * User
       */
      promise = request(/* GraphQL */ `
        {
          user(username: "${second}") {
            name
            username
            description
            # coverPhoto
          }
        }
      `).then(res => {
        const { user } = res.data;
        return generateMetaInfo({
          type: 'user',
          data: {
            username: user.username,
            description: user.description,
            name: user.name,
          },
        });
      });
      break;
    }
    default: {
      const isBlacklisted = PATH_BLACKLIST.includes(first);
      if (second && !isBlacklisted) {
        /**
         * Channel
         */
        promise = request(/* GraphQL */ `
          {
            channel(channelSlug: "${second}", communitySlug: "${first}") {
              name
              description
              isPrivate
              community {
                name
              }
            }
          }
        `).then(res => {
          const { channel, channel: { community } } = res.data;
          return generateMetaInfo({
            type: 'channel',
            data: {
              name: channel.name,
              description: channel.description,
              communityName: community.name,
              private: channel.isPrivate,
            },
          });
        });
      } else if (!isBlacklisted) {
        /**
         * Community
         */
        promise = request(/* GraphQL */ `
          {
            community(slug: "${first}") {
              name
              description
            }
          }
        `).then(res => {
          const { community } = res.data;
          return generateMetaInfo({
            type: 'community',
            data: {
              name: community.name,
              description: community.description,
            },
          });
        });
      }
      break;
    }
  }

  if (!promise) return Promise.resolve(generateMetaInfo());

  return promise.catch(err => {
    console.log(`⚠️ Failed to load metadata for ${url}! ⚠️`);
    console.log(err);
    return generateMetaInfo();
  });
};
