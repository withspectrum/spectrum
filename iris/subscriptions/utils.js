// @flow
import { getUserPermissionsInChannel } from '../models/usersChannels';
import { getMembersInDirectMessageThread } from '../models/usersDirectMessageThreads';
import { getChannels } from '../models/channel';
import type { DBThread } from '../models/thread';
import type { GraphQLContext } from '../';

// Check that a user can view a channel
export const userCanViewChannel = async (channelId: string, userId: string) => {
  const [channel] = await getChannels([channelId]);
  // Everybody can view public channels
  if (!channel.isPrivate) return true;

  if (!userId) return false;
  return await userIsMemberOfChannel(channelId, userId);
};

export const userIsMemberOfChannel = async (
  channelId: string,
  userId: string
) => {
  if (!userId) return false;
  const { isMember } = await getUserPermissionsInChannel(channelId, userId);
  return isMember;
};

export const userCanViewDirectMessageThread = async (
  threadId: string,
  userId: string
) => {
  const members = await getMembersInDirectMessageThread(threadId);
  if (!members) return false;
  return members.some(member => member.userId === userId);
};
