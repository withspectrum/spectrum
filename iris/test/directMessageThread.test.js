// @flow
import { request } from './utils';
import data from 'shared/testing/data';
import { toPlainText, toState } from 'shared/draft-utils';

const messageToPlainText = message =>
  toPlainText(toState(JSON.parse(message.content.body)));

const context = {
  user: data.users.find(({ username }) => username === 'mxstbr'),
};

// All the messages in the test DM thread ordered by time, desc
const messages = data.messages
  .filter(({ threadId }) => threadId === 'first-dm-thread-asdf123')
  .sort((a, b) => a.timestamp - b.timestamp);

it('should fetch a directMessageThread', async () => {
  const query = /* GraphQL */ `
    {
      directMessageThread(id: "first-dm-thread-asdf123") {
        id
        threadLastActive
      }
    }
  `;

  expect.assertions(1);
  const result = await request(query, { context });

  expect(result).toMatchSnapshot();
});

describe('messageConnection', () => {
  it('should fetch a directMessageThreads messages', async () => {
    const query = /* GraphQL */ `
      {
        directMessageThread(id: "first-dm-thread-asdf123") {
          messageConnection(first: 100) {
            edges {
              node {
                id
              }
            }
          }
        }
      }
    `;

    expect.assertions(1);
    const result = await request(query, { context });

    expect(result).toMatchSnapshot();
  });

  it('should fetch the last message first', async () => {
    const query = /* GraphQL */ `
      {
        directMessageThread(id: "first-dm-thread-asdf123") {
          messageConnection(first: 1) {
            pageInfo {
              hasNextPage
              hasPreviousPage
            }
            edges {
              node {
                id
                content {
                  body
                }
              }
            }
          }
        }
      }
    `;

    expect.assertions(2);
    const result = await request(query, { context });

    expect(result.data.directMessageThread.messageConnection.pageInfo).toEqual({
      hasNextPage: true,
      hasPreviousPage: false,
    });

    expect(
      messageToPlainText(
        result.data.directMessageThread.messageConnection.edges[0].node
      )
    ).toEqual(messageToPlainText(messages[messages.length - 1]));
  });

  it('should fetch the second to last message next', async () => {
    // Get the first message, same as above
    const query = /* GraphQL */ `
      {
        directMessageThread(id: "first-dm-thread-asdf123") {
          messageConnection(first: 1) {
            pageInfo {
              hasNextPage
              hasPreviousPage
            }
            edges {
              cursor
            }
          }
        }
      }
    `;

    expect.assertions(3);
    const result = await request(query, { context });

    // Get the cursor of the first message
    // Make sure pageInfo is calculated correctly
    expect(result.data.directMessageThread.messageConnection.pageInfo).toEqual({
      hasNextPage: true,
      hasPreviousPage: false,
    });

    const cursor =
      result.data.directMessageThread.messageConnection.edges[0].cursor;

    // Generate a query of the first message after the cursor of the last message
    const nextQuery = /* GraphQL */ `
      {
        directMessageThread(id: "first-dm-thread-asdf123") {
          messageConnection(first: 1, after: "${cursor}") {
            pageInfo {
              hasNextPage
              hasPreviousPage
            }
            edges {
              node {
                content {
                  body
                }
              }
            }
          }
        }
      }
    `;

    const nextResult = await request(nextQuery, { context });

    expect(
      nextResult.data.directMessageThread.messageConnection.pageInfo
    ).toEqual({
      hasNextPage: true,
      hasPreviousPage: true,
    });

    expect(
      messageToPlainText(
        nextResult.data.directMessageThread.messageConnection.edges[0].node
      )
    ).toEqual(messageToPlainText(messages[messages.length - 2]));
  });
});
