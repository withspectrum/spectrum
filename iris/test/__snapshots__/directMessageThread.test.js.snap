// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messageConnection should fetch a directMessageThreads messages 1`] = `
Object {
  "data": Object {
    "directMessageThread": Object {
      "messageConnection": Object {
        "edges": Array [
          Object {
            "node": Object {
              "id": "9356f7f1-4be0-4230-8f79-71a9f97f7a1f",
            },
          },
          Object {
            "node": Object {
              "id": "77e9619d-bea2-443a-bb9c-4b2957980644",
            },
          },
          Object {
            "node": Object {
              "id": "965830bb-e79d-4cec-acfa-91f67b590865",
            },
          },
          Object {
            "node": Object {
              "id": "2017f510-fe85-4de2-b5b2-6324567a6871",
            },
          },
          Object {
            "node": Object {
              "id": "b546f27e-d842-4afe-900d-ad26ee60ca03",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`should fetch a directMessageThread 1`] = `
Object {
  "data": Object {
    "directMessageThread": Object {
      "id": "first-dm-thread-asdf123",
      "threadLastActive": "2016-12-31T23:00:00.000Z",
    },
  },
}
`;
